### ISSUE TITLE
Enhancement: ParsedSignature to enable TypeVar expansion using signature namespace 

### ISSUE THREAD

**Issue by harryle95** [litestar-org/litestar#3206]
### Summary

ParsedSignature uses `typing.get_type_hints` which handles forward referencing but no TypeVar expansion. For instance, 

```Python
from typing import TypeVar, Generic, get_type_hints

T = TypeVar("T")

class Foo(Generic[T]):
  def bar(self, data: T) -> T:
    pass 

genericAliasFoo = Foo[str]
print(get_type_hints(
  genericAliasFoo.bar, 
  globalns={"T": str}, 
  localns=None)) # gives {'data': ~T, 'return': ~T} 
```
This makes it difficult to write type generic handlers as there is no way of expanding TypeVar. Both `pydantic` and `mypy` use some forms of TypeVar expansion for the same scenario. A naive implementation can be as simple as this: 

```Python
class ParsedSignature:
  @classmethod
  def from_fn(cls, fn: AnyCallable, signature_namespace: dict[str, Any]) -> Self:
      signature = Signature.from_callable(fn)
      fn_type_hints = get_fn_type_hints(fn, namespace=signature_namespace)

      # Expand type var
      for param, value in fn_type_hints.items():
        if value.__name__ in signature_namespace and isinstance(value, TypeVar):
          fn_type_hints[param] = signature_namespace[value]
      return cls.from_signature(signature, fn_type_hints)
```

### Basic Example

As an example, to build a generic controller, this is what works currently (from #1311 and #2162)

```Python
class GenericController(Controller, Generic[T]):
    model_type: type[T]

    def __class_getitem__(cls, model_type: type[T]) -> type:
        return type(
            f"Controller[{model_type.__name__}]", (cls,), {"model_type": model_type}
        )

    def __init__(self, owner: Router):
        super().__init__(owner)
        self.signature_namespace[T.__name__] = self.model_type


class BaseController(GenericController[T]):
    @post()
    async def create(self, data: T.__name__) -> T.__name__:
        return data
```

Note how satanic the post handler looks. This works because `T.__name__` is resolved using `ForwardRef` in `get_type_hints`. Under the new proposal, it is possible to do this instead: 

```Python
class BaseController(GenericController[T]):
    @post()
    async def create(self, data: T) -> T:
        return data
```

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3206">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3206/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/3206/pledge.svg">
</picture>
</a>

**Comment by peterschutt** [comment]
We have https://github.com/litestar-org/litestar/blob/ef9dbb7a0cc4c7031dbf9a7bff02719d8a65a008/litestar/utils/typing.py#L233-L262

**Comment by haryle** [comment]
> We have
> 
> https://github.com/litestar-org/litestar/blob/ef9dbb7a0cc4c7031dbf9a7bff02719d8a65a008/litestar/utils/typing.py#L233-L262

I see, but `fn_type_hints` no longer contains the expanded type information after line 214. In the previous example, `fn_type_hint = {'data': ~T, 'return': ~T}`

https://github.com/litestar-org/litestar/blob/ef9dbb7a0cc4c7031dbf9a7bff02719d8a65a008/litestar/utils/signature.py#L214

**Comment by peterschutt** [comment]
Not sure what you mean. I'm just pointing out that we have a utility that does type `get_type_hints()` with type var expansion.

**Comment by haryle** [comment]
Sorry, I wasn't clear. Correct me if I am wrong, but I don't think that utility method gets called in the `ParsedSignature.from_fn` call stack.

**Comment by peterschutt** [comment]
Obviously it was me who was unclear. I wasn't saying it did - just pointing out we have the utility you mentioned which maybe we can use. LMK how it goes.

**Comment by haryle** [comment]
Ah thank you. Should I make a PR after adding it in? Sorry I am quite new to contributing to open source project.

**Comment by peterschutt** [comment]
You are more than welcome to make a PR, cheers.

To make it easier to review I'd first suggest creating a minimal example (I'm talking about something we can copy and paste and run without any extra work) that demonstrates the issue against `main`. Then we have some context when reviewing the PR.

**Comment by Alc-Alc** [comment]
Closed by #3242
